### PR TITLE
Add preprocessor pass on visual shader when showing generated code

### LINKED ISSUE #82555: Using preprocessor in visual shader global expression produces error "Tokenizer: unknown character #35"
### Godot version

v4.1.stable.official [970459615]

### System information

Godot v4.1.stable - Windows 10.0.19044 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 3060 Ti (NVIDIA; 30.0.14.7141) - AMD Ryzen 5 5600X 6-Core Processor (12 Threads)

### Issue description

screen

include directive should be parsed by preprocessor in visual shader editor. This issue affects only editor, "show generated shader code" button.

### Steps to reproduce

1. Create shader include file
2. Create visual shader
3. Add global expression node to visual shader with include directive
4. Press "show generated shader code" button. Error will be displayed.

### Minimal reproduction project

[vsshader_include.zip](https://github.com/godotengine/godot/files/12772743/vsshader_include.zip)

### PR DESCRIPTION
**Pull request by 100gold**
Fixes #82555 
Add preprocessor pass and check for errors in visual shader editor.

### CODE REVIEW

**Comment by bitsawer** [comment]
Looks pretty good. I would also add the preprocessor error file name, which makes it easier to find the actual issue if you have complex or nested includes.

Also, looks like the error line number highlight is off by one, needs `- 1`. Updated section of the code would be this:


```cpp
if (err != OK) {
    ERR_FAIL_COND(err_positions.size() == 0);

    String file = err_positions.front()->get().file;
    int err_line = err_positions.front()->get().line;
    Color error_line_color = EDITOR_GET("text_editor/theme/highlighting/mark_color");
    preview_text->set_line_background_color(err_line - 1, error_line_color);
    error_panel->show();

    error_label->set_text("error(" + file + ":" + itos(err_line) + "): " + error_pp);
    shader_error = true;
    return;
}
```

**Comment by 100gold** [comment]
> Looks pretty good. I would also add the preprocessor error file name, which makes it easier to find the actual issue if you have complex or nested includes.
> 
> Also, looks like the error line number highlight is off by one, needs `- 1`. Updated section of the code would be this:
> 
> ...

Thanks! There was another issue with the error line number when there was an error in the included file. Fixed both issues.

File: editor/plugins/visual_shader_editor_plugin.cpp [hunk 1]
```diff
@@ -5083,20 +5084,52 @@ void VisualShaderEditor::_update_preview() {
...
+	}
+
+	ShaderLanguage sl;
+	Error err = sl.compile(preprocessed_code, info);
 	if (err != OK) {
+		int err_line;
+		String err_text;
+		Vector<ShaderLanguage::FilePosition> include_positions = sl.get_include_positions();
+		if (include_positions.size() > 1) {
+			//error is in an include
```
Comment (bitsawer) [inline editor/plugins/visual_shader_editor_plugin.cpp:5120]:
```suggestion
			// Error is in an include.
```

**Review by bitsawer — CHANGES_REQUESTED** [review]
Tested locally and works as expected. One comment should be updated to the new format (capitalized and ends with a dot), but other than that looks good.

As a note to myself and future maintainers, this code is now pretty similar to ShaderTextEditor version with some small differences, so it might be possible to merge them in the future. No need to do that in this PR, however.

**Comment by AThousandShips** [comment]
Your commit seems not to be linked to your GitHub account. See: [Why are my commits linked to the wrong user?](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user) for more info.

**Review by clayjohn — APPROVED** [review]
Seems fine to me. I am not familiar with this code, but I trust the other reviewers

**Review by bitsawer — APPROVED** [review]

**Review by AThousandShips — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks! And congrats for your first merged Godot contribution :tada:

**Pull request merged**